darwin/dataset/remote_dataset.py

```diff
@@ -351,7 +351,24 @@ def pull(
             for error in errors:
                 self.console.print(f"\t - {error}")
 
-            downloaded_file_count = len([f for f in self.local_images_path.rglob("*") if f.is_file()])
+            # Remove images that don't have a corresponding annotation file in the release
```

Comment:
Is this necessary? The underlying `download_all_images_from_annotations` function gets passed the remove_extra flag and is meant to already unlink/delete files that don't have an annotation. If that's not doing it's job and still needs to be cleaned up afterwards then perhaps this is a separate ticket?

Reply:
I actually want to clarify one thing here - Is the intended behaviour of `remove_extra=True` to remove files that don't have a corresponding annotation file, or to remove files that are not part of the release currently being pulled?

Currently, the behaviour is the latter, because we remove all files that don't have a stem matching any stem in the release-specific annotations folder

If the latter is the intended behaviour, it works unless there are duplicate filename stems in the release, because it relies on the `existing_images` dict that I changed to a list (see comment above for why a dict was problematic). If `existing_images` is changed to a list or a set of paths where there are duplicate stems in the release, I can't think of a way to get a guaranteed mapping between annotation file and corresponding image/video etc. other than by inspecting the `local_path` field inside `source_files` for each annotation file. However, this isn't populated in annotation files that are part of the current pull **until** the download function is actually finished (i.e. exhaust_generator() is run). As far as I can tell, this means that the removal of extra files has to take place after this  (I chose to put it inside `remote_dataset.py`, but it could easily be put into a function)

Would this be good to put into a new ticket? There was already an issue with `remove_extra` before my changes (problematic `existing_images` dict), but my changes broke it more

Reply:
I don't mind this here, but creating an extra ticket and progressing it with this merge would be good for tracking.

Reply:
Have made changes to ensure that `remove_extra` functions exactly as it did before the ticket, will raise it separately

Reply:
@rslota Would you be able to clarify on my comment above? In working on this issue, I discovered a problem with the remove_extra flag, and neither Nathan nor I are able to tell what it's intended behaviour is

I'll create a new ticket to work on it

Reply:
Sorry for late reply, I missed the notification. 
`remove_extra` afaik is here to remove any image files that is already downloaded that would not get overridden by the current release being pulled. In other words, it should work the same as wiping image directory before pulling the release, however, it can't be implemented this way, in case we have issues downloading "new" images, and customer would ended up with empty target directory.